### PR TITLE
Add GLPN

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR implements [GLPN](https://arxiv.org/abs/2201.07436), Global Local Path Networks, for state-of-the-art depth estimation (number 2 on paperswithcode for KITTI and NYUv2 benchmarks).

To do:

- [x] transfer weights to author's username
- [x] add tests for GLPNFeatureExtractor
- [x] rename `in_index` to something better
- [x] define `DepthEstimatorOutput`

### CODE REVIEW

File: src/transformers/models/glpn/configuration_glpn.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        drop_path_rate (`float`, *optional*, defaults to 0.1):
+            The dropout probability for stochastic depth, used in the blocks of the Transformer encoder.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-6):
+            The epsilon used by the layer normalization layers.
+        max_depth (`int`, *optional*, defaults to 10):
+            The maximum depth of the decoder.
+        in_index (`int`, *optional*, defaults to -1):
+            The index of the features to use in the head.
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/configuration_glpn.py:82]: To be updated

File: src/transformers/models/glpn/feature_extraction_glpn.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        self.do_resize = do_resize
+        self.size_divisor = size_divisor
+        self.resample = resample
+        self.do_rescale = do_rescale
+
+    def _resize(self, image, size_divisor, resample):
+        if not isinstance(image, Image.Image):
+            image = self.to_pil_image(image)
+
+        w, h = image.size
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/feature_extraction_glpn.py:64]:
```suggestion
width, height = image.size
```

File: src/transformers/models/glpn/modeling_glpn.py [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+    "nielsr/glpn-kitti",
+    # See all GLPN models at https://huggingface.co/models?filter=glpn
+]
+
+
+# Copied from transformers.models.segformer.modeling_segformer.to_2tuple
+def to_2tuple(x):
+    if isinstance(x, collections.abc.Iterable):
+        return x
+    return (x, x)
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:65]: can we finally remove this guy?

File: src/transformers/models/glpn/modeling_glpn.py [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+            hidden_size,
+            kernel_size=patch_size,
+            stride=stride,
+            padding=(patch_size[0] // 2, patch_size[1] // 2),
+        )
+
+        self.layer_norm = nn.LayerNorm(hidden_size)
+
+    def forward(self, pixel_values):
+        x = self.proj(pixel_values)
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:120]:
```suggestion
hidden_state = self.proj(pixel_values)
```

same below

File: src/transformers/models/glpn/modeling_glpn.py [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+        _, _, height, width = x.shape
+        x = x.flatten(2).transpose(1, 2)
+        x = self.layer_norm(x)
+        return x, height, width
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:124]: maybe return them in tuple (cleaner)

```suggestion
return hidden_state, (height, width)
```

File: src/transformers/models/glpn/modeling_glpn.py [hunk 6]
```diff
@@ -0,0 +1,736 @@
...
+    def transpose_for_scores(self, x):
+        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
+        x = x.view(*new_x_shape)
+        return x.permute(0, 2, 1, 3)
+
+    def forward(
+        self,
+        hidden_states,
+        height,
+        width,
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:163]: Very personal comment, I think is cleaner if you encapsulate `(height, width)` in something like `spatial_dimensions`

File: src/transformers/models/glpn/modeling_glpn.py [hunk 7]
```diff
@@ -0,0 +1,736 @@
...
+        output_attentions=False,
+    ):
+        query_layer = self.transpose_for_scores(self.query(hidden_states))
+
+        if self.sr_ratio > 1:
+            batch_size, seq_len, num_channels = hidden_states.shape
+            hidden_states = hidden_states.permute(0, 2, 1).reshape(batch_size, num_channels, height, width)
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:170]: comment about the rearrange?

File: src/transformers/models/glpn/modeling_glpn.py [hunk 8]
```diff
@@ -0,0 +1,736 @@
...
+            hidden_states = self.sr(hidden_states)
+            hidden_states = hidden_states.reshape(batch_size, num_channels, -1).permute(0, 2, 1)
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:172]: comment about the rearrange?

File: src/transformers/models/glpn/modeling_glpn.py [hunk 9]
```diff
@@ -0,0 +1,736 @@
...
+        # Normalize the attention scores to probabilities.
+        attention_probs = nn.functional.softmax(attention_scores, dim=-1)
+
+        # This is actually dropping out entire tokens to attend to, which might
+        # seem a bit unusual, but is taken from the original Transformer paper.
+        attention_probs = self.dropout(attention_probs)
+
+        context_layer = torch.matmul(attention_probs, value_layer)
+
+        context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:192]: comment about the rearrange?

File: src/transformers/models/glpn/modeling_glpn.py [hunk 10]
```diff
@@ -0,0 +1,736 @@
...
+        return outputs
+
+
+class GLPNEncoder(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+
+        # stochastic depth decay rule
+        dpr = [x.item() for x in torch.linspace(0, config.drop_path_rate, sum(config.depths))]
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:336]: `dpr` -> `drop_path_ decays`

File: src/transformers/models/glpn/modeling_glpn.py [hunk 11]
```diff
@@ -0,0 +1,736 @@
...
+            embeddings.append(
+                GLPNOverlapPatchEmbeddings(
+                    image_size=config.image_size // config.downsampling_rates[i],
+                    patch_size=config.patch_sizes[i],
+                    stride=config.strides[i],
+                    num_channels=config.num_channels if i == 0 else config.hidden_sizes[i - 1],
+                    hidden_size=config.hidden_sizes[i],
+                )
+            )
+        self.patch_embeddings = nn.ModuleList(embeddings)
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:350]: Looks like it can be a `*Stage`

File: src/transformers/models/glpn/modeling_glpn.py [hunk 12]
```diff
@@ -0,0 +1,736 @@
...
+            # second, send embeddings through blocks
+            for i, blk in enumerate(block_layer):
+                layer_outputs = blk(hidden_states, height, width, output_attentions)
+                hidden_states = layer_outputs[0]
+                if output_attentions:
+                    all_self_attentions = all_self_attentions + (layer_outputs[1],)
+            # third, apply layer norm
+            hidden_states = norm_layer(hidden_states)
+            # fourth, optionally reshape back to (batch_size, num_channels, height, width)
+            hidden_states = hidden_states.reshape(batch_size, height, width, -1).permute(0, 3, 1, 2).contiguous()
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:406]: You can create a `GPTStage` that contains one `embedding` one `transformer block (layer!)* and one `norm_layer`

File: src/transformers/models/glpn/modeling_glpn.py [hunk 13]
```diff
@@ -0,0 +1,736 @@
...
+        self.skip_conv2 = nn.Conv2d(in_channels=config.hidden_sizes[-3], out_channels=out_channels, kernel_size=1)
+
+        self.up = nn.Upsample(scale_factor=2, mode="bilinear", align_corners=False)
+
+        self.fusion1 = GLPNSelectiveFeatureFusion(out_channels)
+        self.fusion2 = GLPNSelectiveFeatureFusion(out_channels)
+        self.fusion3 = GLPNSelectiveFeatureFusion(out_channels)
+
+    def forward(self, encoder_hidden_states):
+        x_1, x_2, x_3, x_4 = encoder_hidden_states
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:558]: just iterate them!

File: src/transformers/models/glpn/modeling_glpn.py [hunk 14]
```diff
@@ -0,0 +1,736 @@
...
+        output.append(out)
+
+        return output
+
+
+class GLPNSelectiveFeatureFusion(nn.Module):
+    def __init__(self, in_channel=64):
+        super().__init__()
+
+        self.conv1 = nn.Sequential(
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:588]: `sel.convolution1`

Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:588]: `self.convolution_layer1`?

File: src/transformers/models/glpn/modeling_glpn.py [hunk 15]
```diff
@@ -0,0 +1,736 @@
...
+
+        out = x_local * attn[:, 0, :, :].unsqueeze(1) + x_global * attn[:, 1, :, :].unsqueeze(1)
+
+        return out
+
+
+class SiLogLoss(nn.Module):
+    def __init__(self, lambd=0.5):
+        super().__init__()
+        self.lambd = lambd
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:619]: Let's add a docstring with the formula of the loss

File: src/transformers/models/glpn/modeling_glpn.py [hunk 16]
```diff
@@ -0,0 +1,736 @@
...
+        features = config.hidden_sizes[0]
+        self.head = nn.Sequential(
+            nn.Conv2d(features, features, kernel_size=3, stride=1, padding=1),
+            nn.ReLU(inplace=False),
+            nn.Conv2d(features, 1, kernel_size=3, stride=1, padding=1),
+        )
+
+    def forward(self, hidden_states: List[torch.Tensor]) -> torch.Tensor:
+        # use last features of the decoder
+        hidden_states = hidden_states[self.config.in_index]
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:644]: cool!

**Review by FrancescoSaverioZuppichini — APPROVED** [review]
Thanks! LGTM! There are comments about variable names and design choice that we may take to simplify the code

File: src/transformers/__init__.py [hunk 17]
```diff
@@ -633,6 +635,7 @@
     _import_structure["modeling_utils"] = ["Conv1D", "PreTrainedModel", "apply_chunking_to_forward", "prune_layer"]
 
     # PyTorch models structure
+
```
Comment (sgugger) [inline src/transformers/__init__.py:638]:
```suggestion
```

File: src/transformers/__init__.py [hunk 18]
```diff
@@ -993,6 +996,15 @@
             "load_tf_weights_in_funnel",
         ]
     )
+    _import_structure["models.glpn"].extend(
+        [
+            "GLPN_PRETRAINED_MODEL_ARCHIVE_LIST",
+            "GLPNForDepthEstimation",
+            "GLPNLayer",
```
Comment (sgugger) [inline src/transformers/__init__.py:1003]:
```suggestion
```

File: src/transformers/__init__.py [hunk 19]
```diff
@@ -3195,6 +3210,13 @@
             FunnelPreTrainedModel,
             load_tf_weights_in_funnel,
         )
+        from .models.glpn import (
+            GLPN_PRETRAINED_MODEL_ARCHIVE_LIST,
+            GLPNForDepthEstimation,
+            GLPNLayer,
```
Comment (sgugger) [inline src/transformers/__init__.py:3216]:
```suggestion
```

File: src/transformers/models/glpn/configuration_glpn.py [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+            Sequence reduction ratios in each encoder block.
+        hidden_sizes (`List[int]`, *optional*, defaults to [32, 64, 160, 256]):
+            Dimension of each of the encoder blocks.
+        downsampling_rates (`List[int]`, *optional*, defaults to [1, 4, 8, 16]):
+            Downsample rate of the image resolution compared to the original image size before each encoder block.
+        patch_sizes (`List[int]`, *optional*, defaults to [7, 3, 3, 3]):
+            Patch size before each encoder block.
+        strides (`List[int]`, *optional*, defaults to [4, 2, 2, 2]):
+            Stride before each encoder block.
+        num_attention_heads (`List[int]`, *optional*, defaults to [1, 2, 4, 8]):
```
Comment (sgugger) [inline src/transformers/models/glpn/configuration_glpn.py:59]: Lists are objects so should be between \` \` in all those defaults.

File: src/transformers/models/glpn/configuration_glpn.py [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+            Number of attention heads for each attention layer in each block of the Transformer encoder.
+        mlp_ratios (`List[int]`, *optional*, defaults to [4, 4, 4, 4]):
```
Comment (sgugger) [inline src/transformers/models/glpn/configuration_glpn.py:61]:
```suggestion
        mlp_ratios (`List[int]`, *optional*, defaults to `[4, 4, 4, 4]`):
```

File: src/transformers/models/glpn/modeling_glpn.py [hunk 22]
```diff
@@ -0,0 +1,736 @@
...
+                    stride=config.strides[i],
+                    num_channels=config.num_channels if i == 0 else config.hidden_sizes[i - 1],
+                    hidden_size=config.hidden_sizes[i],
+                )
+            )
+        self.patch_embeddings = nn.ModuleList(embeddings)
+
+        # Transformer blocks
+        blocks = []
+        cur = 0
```
Comment (sgugger) [inline src/transformers/models/glpn/modeling_glpn.py:354]: Can we use the full name of what `cur` stands for (`current`?)

File: src/transformers/models/glpn/modeling_glpn.py [hunk 23]
```diff
@@ -0,0 +1,736 @@
...
+            if output_hidden_states:
+                all_hidden_states = all_hidden_states + (hidden_states,)
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, all_hidden_states, all_self_attentions] if v is not None)
+        return BaseModelOutput(
+            last_hidden_state=hidden_states,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attentions,
+        )
```
Comment (sgugger) [inline src/transformers/models/glpn/modeling_glpn.py:416]:
```suggestion
        return BaseModelOutput(
            last_hidden_state=hidden_states, hidden_states=all_hidden_states, attentions=all_self_attentions
        )
```

File: src/transformers/models/glpn/modeling_glpn.py [hunk 24]
```diff
@@ -0,0 +1,736 @@
...
+        sequence_output = encoder_outputs[0]
+
+        if not return_dict:
+            return (sequence_output,) + encoder_outputs[1:]
+
+        return BaseModelOutput(
+            last_hidden_state=sequence_output,
+            hidden_states=encoder_outputs.hidden_states,
+            attentions=encoder_outputs.attentions,
+        )
```
Comment (sgugger) [inline src/transformers/models/glpn/modeling_glpn.py:538]:
```suggestion
        return BaseModelOutput(
            last_hidden_state=hidden_states, hidden_states=all_hidden_states, attentions=all_self_attentions
        )
```

File: tests/glpn/test_modeling_glpn.py [hunk 25]
```diff
@@ -0,0 +1,360 @@
...
+class GLPNModelTest(ModelTesterMixin, unittest.TestCase):
+
+    all_model_classes = (
+        (
+            GLPNModel,
+            GLPNForDepthEstimation,
+        )
+        if is_torch_available()
+        else ()
+    )
```
Comment (sgugger) [inline tests/glpn/test_modeling_glpn.py:152]:
```suggestion
    all_model_classes = (GLPNModel, GLPNForDepthEstimation) if is_torch_available() else ()
```

**Review by sgugger — APPROVED** [review]
Thanks for adding this new model!

File: src/transformers/models/glpn/modeling_glpn.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/models/glpn/modeling_glpn.py:65]: In a separate PR, yes

File: src/transformers/models/glpn/modeling_glpn.py [hunk 26]
```diff
@@ -0,0 +1,736 @@
...
+            stride=stride,
+            padding=(patch_size[0] // 2, patch_size[1] // 2),
+        )
+
+        self.layer_norm = nn.LayerNorm(hidden_size)
+
+    def forward(self, pixel_values):
+        x = self.proj(pixel_values)
+        _, _, height, width = x.shape
+        x = x.flatten(2).transpose(1, 2)
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/glpn/modeling_glpn.py:122]: Separate PR

File: src/transformers/models/glpn/modeling_glpn.py (hunk 6 above)
Comment (NielsRogge) [reply src/transformers/models/glpn/modeling_glpn.py:163]: Separate PR

File: src/transformers/models/glpn/modeling_glpn.py (hunk 12 above)
Comment (NielsRogge) [reply src/transformers/models/glpn/modeling_glpn.py:406]: I can't because of copied from

**Comment by NielsRogge** [comment]
Ok I've addressed all comments. Waiting for the author to respond such that I can transfer the weights to his name.

File: src/transformers/models/glpn/configuration_glpn.py [hunk 27]
```diff
@@ -80,7 +80,7 @@ class GLPNConfig(PretrainedConfig):
             The dimension of the decoder.
         max_depth (`int`, *optional*, defaults to 10):
             The maximum depth of the decoder.
-        in_index (`int`, *optional*, defaults to -1):
+        head_in_index (`int`, *optional*, defaults to -1):
             The index of the features to use in the head.
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/configuration_glpn.py:84]:
```suggestion
The index of the hidden state (feature) from the backbone to use in the head.
```
Maybe a little more descriptive?

File: src/transformers/models/glpn/modeling_glpn.py [hunk 28]
```diff
@@ -0,0 +1,746 @@
...
+        self.value = nn.Linear(self.hidden_size, self.all_head_size)
+
+        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
+
+        self.sr_ratio = sr_ratio
+        if sr_ratio > 1:
+            self.sr = nn.Conv2d(hidden_size, hidden_size, kernel_size=sr_ratio, stride=sr_ratio)
+            self.layer_norm = nn.LayerNorm(hidden_size)
+
+    def transpose_for_scores(self, x):
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:154]: No one name variables :)

File: src/transformers/models/glpn/modeling_glpn.py [hunk 29]
```diff
@@ -0,0 +1,746 @@
...
+        attention_output = self.output(self_outputs[0], hidden_states)
+        outputs = (attention_output,) + self_outputs[1:]  # add attentions if we output them
+        return outputs
+
+
+# Copied from transformers.models.segformer.modeling_segformer.SegformerDWConv
+class GLPNDWConv(nn.Module):
+    def __init__(self, dim=768):
+        super().__init__()
+        self.dwconv = nn.Conv2d(dim, dim, 3, 1, 1, bias=True, groups=dim)
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:254]:
```suggestion
self.depth_wise_convolution = nn.Conv2d(dim, dim, 3, 1, 1, bias=True, groups=dim)
```

File: src/transformers/models/glpn/modeling_glpn.py [hunk 30]
```diff
@@ -0,0 +1,746 @@
...
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x_local, x_global):
+        x = torch.cat((x_local, x_global), dim=1)
+        x = self.convolutional_layer1(x)
+        x = self.convolutional_layer2(x)
+        x = self.convolutional_layer3(x)
+        attn = self.sigmoid(x)
+
+        out = x_local * attn[:, 0, :, :].unsqueeze(1) + x_global * attn[:, 1, :, :].unsqueeze(1)
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:570]: add a comment here, so we know what is going on :)

File: src/transformers/models/glpn/modeling_glpn.py [hunk 31]
```diff
@@ -0,0 +1,746 @@
...
+        )
+
+        self.convolutional_layer3 = nn.Conv2d(
+            in_channels=int(in_channel / 2), out_channels=2, kernel_size=3, stride=1, padding=1
+        )
+
+        self.sigmoid = nn.Sigmoid()
+
+    def forward(self, x_local, x_global):
+        x = torch.cat((x_local, x_global), dim=1)
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:564]: no one name variables

File: src/transformers/models/glpn/modeling_glpn.py [hunk 32]
```diff
@@ -0,0 +1,746 @@
...
+
+        stage_hidden_states[-1] = self.final_upsample(stage_hidden_state)
+
+        return stage_hidden_states
+
+
+class SiLogLoss(nn.Module):
+    """
+    Implements the Scale-invariant log scale loss [Eigen et al., 2014](https://arxiv.org/abs/1406.2283).
+    """
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:625]: Do you think we can add the formula?

File: src/transformers/models/glpn/modeling_glpn.py [hunk 33]
```diff
@@ -0,0 +1,746 @@
...
+        return loss
+
+
+class GLPNDepthEstimationHead(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+
+        self.config = config
+
+        features = config.hidden_sizes[0]
```
Comment (FrancescoSaverioZuppichini) [inline src/transformers/models/glpn/modeling_glpn.py:645]: we don't use `features` usually, what about `channels`?

**Review by FrancescoSaverioZuppichini — APPROVED** [review]
A couple of comments :)

File: src/transformers/models/glpn/modeling_glpn.py (hunk 28 above)
Comment (NielsRogge) [reply src/transformers/models/glpn/modeling_glpn.py:154]: Can't change this, it's used in 60+ models

File: src/transformers/models/glpn/modeling_glpn.py (hunk 29 above)
Comment (NielsRogge) [reply src/transformers/models/glpn/modeling_glpn.py:254]: Sorry, it's copied from SegFormer :(

**Pull request merged**